docs/specs/priority-queue.md

```diff
@@ -1,281 +1,194 @@
-    public PriorityQueue(IComparer<TPriority> comparer);
-    public PriorityQueue(IEnumerable<(TElement, TPriority)> collection);
-    public PriorityQueue(IEnumerable<(TElement, TPriority)> collection, IComparer<TPriority> comparer);
-
-    public IComparer<TPriority> Comparer { get; }
+    // inherited from IDictionary<TKey, TPriority>
     public int Count { get; }
-
-    public bool IsEmpty { get; }
-    public void Clear();
-    public bool Contains(TElement element);
-    public bool Contains(TElement element, out TPriority priority);
-
-    public void Enqueue(TElement element, TPriority priority);
-
-    public (TElement element, TPriority priority) Peek();
-    public bool TryPeek(out TElement element, out TPriority priority);
-    public bool TryPeek(out TElement element);
-
-    public (TElement element, TPriority priority) Dequeue();
-    public bool TryDequeue(out TElement element, out TPriority priority);
-    public bool TryDequeue(out TElement element);
-
-    public bool Remove(TElement element);
-    public bool Remove(TElement element, out TPriority priority);
-
-    public IEnumerator<(TElement element, TPriority priority)> GetEnumerator();
-    IEnumerator IEnumerable.GetEnumerator();
-
-    public IEnumerable<TElement> Elements { get; }
-    public IEnumerable<TPriority> Priorities { get; }
-
-    public struct Enumerator : IEnumerator<(TElement element, TPriority priority)>
+    public bool IsReadOnly { get; }
+    public TPriority Item[TKey] { get; set; } // get or update priorities, can also be used to add keys and their priorities to the queue
+    public ICollection<TKey> Keys { get; } // the keys, not necessarily in sorted order
+    public ICollection<TKPriority> Values { get; } // the priorities, not necessarily in sorted order
+
+    // inherited from IDictionary<TKey, TPriority>
+    public void Add(TKey key, TPriority priority); // throws ArgumentException if the key was already in the collection
+    public void Clear(); 
+    public bool ContainsKey(TKey key);
+    public bool Remove(TKey key); // Returns true if object was successfully removed.
+    public bool TryGetValue(TKey key, out TPriority priority);
+
+    // inherited from ICollection<KeyValuePair<TKey, TPriority>>
+    public void Add(KeyValuePair<TKey, TPriority> item);
+    public bool Contains(KeyValuePair<TKey, TPriority> item);
+    public void CopyTo(KeyValuePair<TKey key, TPriority>[] array, int arrayIndex);
+    public IEnumerator<KeyValuePair<TKey key, TPriority priority>> GetEnumerator();
+    public bool Remove(KeyValuePair<TKey, TPriority> item);
+
+    // new
+    public TKey Dequeue();  // throws InvalidOperationException if the queue is empty (like Queue<>)
+    public void Enqueue(TKey key, TPriority priority); // an alias for IDictionary<TKey,TPriority>.Add()
+    public TKey Peek(TKey key); // throws InvalidOperationException if the queue is empty (like Queue<>)
+    public bool TryDequeue(out TKey key);
```

Comment:
Instead of having two overrides for TryDequeue should we just have one that returns a `KeyValuePair`?

Reply:
Sure, or thinking of C# 7 and out vars + ignored out parameters, we could choose to go the two out-param route... I'm not sure which is ideally preferred, but personally I think we want this to be a bool return type, so maybe the two out parameters route is best?

Reply:
Yes I think the two out parameters would be fine as well. As with that route people will have access to both the key and priority after calling the method, without having to access properties inside the KeyValuePair object.